### PR TITLE
ViT UNETR SSL

### PR DESCRIPTION
**Pull request by valosekj**
Just a draft PR to allow commit suggestions

### CODE REVIEW

File: vit_unetr_ssl/transforms.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+            # Create copies of items in the dictionary under new keys, allowing for the same image to be manipulated
+            # differently in subsequent transformations
+            CopyItemsd(keys=["image"], times=2, names=["gt_image", "image_2"], allow_missing_keys=False),
+
+            # AUGMENTED VIEW 1
+            OneOf(
+                transforms=[
+                    # Randomly drop regions of the image
+                    # 'dropout_holes=True': the dropped regions will be set to zero, introducing regions of no
+                    # information within the image.
```
Comment (naga-karthik) [inline vit_unetr_ssl/transforms.py]: so I think this definition is incorrect -- if `dropout_holes=True` then the `holes` will be dropped out and the regions in the image with dropped holes will be filled with `fill_value` (i.e. they are not always set to zero, depends on `fill_value`)

File: vit_unetr_ssl/transforms.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+                    RandCoarseDropoutd(
+                        keys=["image"],
+                        prob=1.0,
+                        holes=number_of_holes,
+                        spatial_size=roi_size[0] // 4,
+                        dropout_holes=True,
+                        # max_spatial_size=(roi_size[0]//4, roi_size[1]//4, roi_size[2]//4)
+                    ),
+                    # 'dropout_holes=False': the areas inside the holes will be filled with random noise
```
Comment (naga-karthik) [inline vit_unetr_ssl/transforms.py:85]: Likewise, this should be clarified as well! `dropout_holes=False` is just the opposite -- the `holes` will be kept and the rest of the image is dropped out 

Remember, we saw a full images with only the SC in the middel and the rest of the image was noise? So, I guess that the result of 2 successive transforms: (1) `dropout_holes=False` which drops out the entire image (except the holes) and fills the values with 0 (because the default value for `fill_value` is `0`) and (2) `RandCoarseShuffle` shuffles the pixels throughout the image.

File: vit_unetr_ssl/transforms.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+                           magnitude_range=(25., 35.)),
+            RandSimulateLowResolutiond(keys=["image"], zoom_range=(0.5, 1.0), prob=0.25),
+            RandAdjustContrastd(keys=["image"], gamma=(0.5, 3.), prob=0.5),  # this is monai's RandomGamma
+            RandBiasFieldd(keys=["image"], coeff_range=(0.0, 0.5), degree=3, prob=0.3),
+            RandGaussianNoised(keys=["image"], mean=0.0, std=0.1, prob=0.1),
+            RandGaussianSmoothd(keys=["image"], sigma_x=(0., 2.), sigma_y=(0., 2.), sigma_z=(0., 2.0),
+                                prob=0.3),
+            RandScaleIntensityd(keys=["image"], factors=(-0.25, 1), prob=0.15),
+            # this is nnUNet's BrightnessMultiplicativeTransform
+            RandFlipd(keys=["image", "label"], prob=0.3, ),
```
Comment (naga-karthik) [inline vit_unetr_ssl/transforms.py]: oh man, taking these transforms exactly from `contrast-agnostic` is not a good idea! They heavily deform the image (useful for SC seg but not for lesion seg). We need much simpler transforms for lesion seg. Add my suggestions below

Comment (naga-karthik) [inline vit_unetr_ssl/transforms.py]:
```suggestion
            RandAffined(keys=["image", "label"], mode=(2, 1), prob=0.1,
                        rotate_range=(-20. / 360 * 2. * np.pi, 20. / 360 * 2. * np.pi),
                        # monai expects in radians
                        scale_range=(-0.2, 0.2),
                        translate_range=(-0.1, 0.1)),
            RandSimulateLowResolutiond(keys=["image"], zoom_range=(0.5, 1.0), prob=0.2),
            RandAdjustContrastd(keys=["image"], gamma=(0.5, 3.), prob=0.2),  # this is monai's RandomGamma
            RandBiasFieldd(keys=["image"], coeff_range=(0.0, 0.5), degree=3, prob=0.1),
            RandGaussianSmoothd(keys=["image"], sigma_x=(0., 2.), sigma_y=(0., 2.), sigma_z=(0., 2.0),
                                prob=0.2),
            RandFlipd(keys=["image", "label"], prob=0.2, ),
```

File: vit_unetr_ssl/transforms.py (hunk 1 above)
Comment (valosekj) [reply vit_unetr_ssl/transforms.py]: Good catch! You are right!

From [RandCoarseDropoutd](https://docs.monai.io/en/latest/transforms.html#randcoarsedropoutd) documentation:

> dropout_holes – if True, dropout the regions of holes and fill value, if False, keep the holes and dropout the outside and fill value. default to True.

Since, I do not specify `fill_value`, `None` is used, meaning:

> if None, will compute the min and max value of input image then randomly select value to fill, default to None.

I further debugged and plotted the transforms. It confirms your findings! 

image

File: vit_unetr_ssl/transforms.py (hunk 2 above)
Comment (valosekj) [reply vit_unetr_ssl/transforms.py:85]: Again, good catch! You are right. See [my comment above](https://github.com/ivadomed/model-seg-dcm/pull/10#discussion_r1527289602). I'll clarify the comments in future commits.

File: vit_unetr_ssl/transforms.py (hunk 3 above)
Comment (valosekj) [reply vit_unetr_ssl/transforms.py]: Agree. Indeed, I just copy-pasted the transforms. But note that these transforms are for the fine-tuned model, which can theoretically be any type of segmentation (SC, lesions, ...). But agree, I will accept your suggestions! Thanks!

Comment (valosekj) [reply vit_unetr_ssl/transforms.py]: Committed in https://github.com/ivadomed/model-seg-dcm/pull/10/commits/a4182b5f91aa8693e617ae933fde3856fc8a2ad1

File: vit_unetr_ssl/transforms.py (hunk 1 above)
Comment (naga-karthik) [reply vit_unetr_ssl/transforms.py]: So regarding this -- I think it's best to set `fill_value=0`. Reason -- when the dropped out holes are randomly filled with values, then this makes it difficult for the model to learn (i.e. it is trying to learn that "there is a patch of random values, I have to figure out how to understand what's behind them"). 
INSTEAD, if we set `fill_value=0` then we're making the task simpler -- the models sees only black regions and it knows that these patches are uniform (i.e. filled only with zeros) instead of seeing that there are `n` holes and each hole has different values of pixels)

Bottomline -- set `fill_value=0` to make the training objective easier

Comment (valosekj) [reply vit_unetr_ssl/transforms.py]: Thanks for this suggestion! Makes sense! Trying `fill_value=0` now (https://github.com/ivadomed/model-seg-dcm/pull/10/commits/bdb2ac155cb53b64423d7b7cb8e416492629707e).

image

File: vit_unetr_ssl/transforms.py (hunk 2 above)
Comment (valosekj) [reply vit_unetr_ssl/transforms.py:85]: Fixed in https://github.com/ivadomed/model-seg-dcm/pull/10/commits/7f2734084cdb1e0cdaa4e4890992351e0873653d and https://github.com/ivadomed/model-seg-dcm/pull/10/commits/bdb2ac155cb53b64423d7b7cb8e416492629707e

File: vit_unetr_ssl/transforms.py (hunk 1 above)
Comment (naga-karthik) [reply vit_unetr_ssl/transforms.py]: Thanks for the figures! maybe you should keep both `dropout_holes=False` and `dropout_holes=True` but with `fill_value=0`.

Comment (valosekj) [reply vit_unetr_ssl/transforms.py]: Good point! I'm now showing both `dropout_holes=False` and `dropout_holes=True` for both `fill_value=None` and `fill_value=0`:

image

Comment (naga-karthik) [reply vit_unetr_ssl/transforms.py]: so we can go ahead with the `fill_value=0` set of transforms? what do you think?

Comment (valosekj) [reply vit_unetr_ssl/transforms.py]: Yes! I pushed `fill_value=0` in https://github.com/ivadomed/model-seg-dcm/pull/10/commits/bdb2ac155cb53b64423d7b7cb8e416492629707e. The model is currently training!
